Cleanup cache initialization

benediktibk opened the pull request:
Data and instruction cache initialization is currently a little bit messy. This PR aims at unifying the available config options and providing a more consistent approach on cache activation.

This is a follow up to #66443.

Open is to check where the caches were automatically enabled previously (SystemInit, Hal, ...) and disabling them directly after.
- [ ] soc_rt10xx.c
- [ ] soc_rt11xx.c
- [ ] kv5x/soc.c
- [x] same70/soc.c
- [x] samv71/soc.c

carlocaione commented on arch/arc/core/cache.c:
```diff
@@ -218,8 +218,9 @@ int arch_icache_flush_and_invd_range(void *addr, size_t size)
 {
-
-	arch_dcache_enable();
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		arch_dcache_enable();
+	}
```
The problem here is that the architecture must not use `arch_dcache_enable` directly, but instead they have to rely on `sys_cache_data_enable`.

`sys_cache_data_enable` collapse to a NOP when `!CONFIG_CACHE_MANAGEMENT` or `!CONFIG_DCACHE`.

benediktibk replied:
fine for me, will change it

pillo79 replied:
Forgive me if I'm wrong, but that would never enable caches unless both of these settings were enabled, and it makes sense to enable the CPU cache but not the Cache API interface. 
If using `arch_` functions is wrong in this context, perhaps the suggestion should be to use a guarded `cache_data_enable()` instead?

carlocaione replied:
> Forgive me if I'm wrong, but that would never turn on caches unless both of these settings were enabled

Correct, and that is explained in https://docs.zephyrproject.org/latest/hardware/cache/index.html indeed.

`CONFIG_DCACHE` is telling you that a `DCACHE` is present but whether that is actually used or not it is up to `CONFIG_CACHE_MANAGEMENT` option.

> it makes sense to be able to use the CPU cache but not the Cache API interface.

Does it though? Before the cache API was introduced every single architecture was doing something different and custom, the cache API was introduced exactly to put order in the cache management chaos. So the path forward here is just to move to the new cache API, not adding workarounds.

benediktibk replied:
I agree with @pillo79, CONFIG_CACHE_MANAGEMENT might be disabled manually, but still CONFIG_DCACHE enabled. In such a case I would expect not to be able to enable or disable the caches from the application, but still the data cache to be active.

    static ALWAYS_INLINE void sys_cache_data_enable(void)
    {
    #if defined(CONFIG_CACHE_MANAGEMENT) && defined(CONFIG_DCACHE)
        cache_data_enable();
    #endif
    }

I'm rather thinking towards calling cache_data_enable, with IS_ENABLED(CONFIG_DCACHE) around it.

benediktibk replied:
> `CONFIG_DCACHE` is telling you that a `DCACHE` is present but whether that is actually used or not it is up to `CONFIG_CACHE_MANAGEMENT` option.

This is not represented like this in the help for the option:

    This option enables the support for the data cache (d-cache).

There might be a difference being cache available and cache being used/activated.

pillo79 replied:
@carlocaione thanks for the link! Still confused though... quoting from there (emphasis mine):
> `CONFIG_DCACHE` / `CONFIG_ICACHE`: [...] All the code paths related to cache control must be conditionally enabled depending on these symbols. **When the symbol is set the cache is considered enabled and used**.
> These symbols say nothing about the actual API interface exposed to the user. For example a platform using the data cache can enable the `CONFIG_DCACHE` symbol and use some HAL exported function in some platform-specific code to enable and manage the d-cache.

> `CONFIG_CACHE_MANAGEMENT`: this option must be selected when the cache operations are exposed to the user through a standard API (see [Cache API](https://docs.zephyrproject.org/latest/hardware/cache/index.html#cache-api)).

carlocaione replied:
when the cache API was introduced we could not move all the cache implementation over right away, it was just too much work. So where possible we simply used `CONFIG_DCACHE` around the legacy cache operation code and we added `CONFIG_DCACHE` to be set at SoC / platform level to say: "the SoC has a functional DCACHE that is being used by some code".

For legacy code we moved from:
```
  custom_cache_op();
```
to:
```
#ifdef CONFIG_DCACHE
  custom_cache_op();
#endif
```

So, a quick recap is:
- `CONFIG_CPU_HAS_DCACHE`: the CPU has a DCACHE, it can be used or not
- `CONFIG_DCACHE`: the CPU has a DCACHE that is being used somehow (this was done to support legacy cache operations)
- `CONFIG_CACHE_MANAGEMENT`: the CPU has a DCACHE that is being used using the cache API

So, back to this PR. Is this PR correct? It is, but it is simply fixing a legacy usage that it should (and it is) deprecated.

benediktibk replied:
Thanks for the clarification, but I think one version in between is missing.

- `CONFIG_CPU_HAS_DCACHE`: the CPU has a DCACHE, it can be used or not
- `CONFIG_DCACHE`: the CPU has a DCACHE that is being used somehow (this was done to support legacy cache operations)
- `???`: the CPU as DCACHE which is active, but it is not accessible using the cache API
- `CONFIG_CACHE_MANAGEMENT`: the CPU has a DCACHE that is being used using the cache API

I will in the meantime go ahead with moving forward to sys_cache_data_enable as CONFIG_DCACHE seems to be somewhat deprecated.

carlocaione replied:
> the CPU as DCACHE which is active, but it is not accessible using the cache API

Well, that would be `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n`.

That is: you have something managing the DCACHE using some kind of custom API that it is not the zephyr cache API. In this case your code should use the custom / HAL API to manage the cache and now you see why we needed a formal, shared API for that, because we needed to abstract the cache operations away from the actual implementation.

benediktibk replied:
But with `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n` sys_cache_data_enable will turn into NOP. Therefore I won't be able to use the cache management API.
I totally agree with you that having an abstract cache operations API is a good thing to have. I'm just saying that having this API available doesn't necessarily mean that one wants to automatically enable the cache. These are two separate things for me.

carlocaione replied:
> But with CONFIG_DCACHE=y and CONFIG_CACHE_MANAGEMENT=n sys_cache_data_enable will turn into NOP

Yes, exactly.

You can have `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n` in two cases:

1. Your SoC has `sys_cache_data_enabled` implemented but you do not want to use it, this is why you are setting `CONFIG_CACHE_MANAGEMENT=n`, so `sys_cache_data_enable()` is returning `-ENOTSUP`.
2. Your SoC doesn't have `sys_cache_data_enabled` implemented, but to enable the DCACHE it is relying on some custom API, so `sys_cache_data_enable()` is returning `-ENOTSUP` because the API that you want to use is different (custom).

carlocaione reviewed (changes requested):
This is fixing the wrong way a pre-existing issue.

The issue is that the cache APIs must be used and not rely on architecture-specific operations like `SCB_*`. The guards are already included in the APIs, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/cache.h

nandojve commented:
Hi @benediktibk ,

In the case of SAM only:

The initial discussion around SAM cache started here #60425. Later the #61005 solve the issues and let user setup cache as it pleases. The default value is cache disabled and the below examples are for SAMV71 which have both I/D Cache available. User should add below lines on their own board.

I Cache only:
CONFIG_CACHE_MANAGEMENT=y
CONFIG_DCACHE=n

D Cache only:
CONFIG_CACHE_MANAGEMENT=y
CONFIG_ICACHE=n

I/D Cache disabled:
CONFIG_ICACHE=n
CONFIG_DCACHE=n

I/D Cache Enabled:
CONFIG_CACHE_MANAGEMENT=y

IMHO, cache should be user defined since user may want to use ITCM/DTCM for other purposes. So, I'm interested in know how this PR changes the above behavior?

benediktibk commented:
> Hi @benediktibk ,
> 
> In the case of SAM only:
> 
> The initial discussion around SAM cache started here #60425. Later the #61005 solve the issues and let user setup cache as it pleases. The default value is cache disabled and the below examples are for SAMV71 which have both I/D Cache available. User should add below lines on their own board.
> 
> I Cache only: CONFIG_CACHE_MANAGEMENT=y CONFIG_DCACHE=n
> 
> D Cache only: CONFIG_CACHE_MANAGEMENT=y CONFIG_ICACHE=n
> 
> I/D Cache disabled: CONFIG_ICACHE=n CONFIG_DCACHE=n
> 
> I/D Cache Enabled: CONFIG_CACHE_MANAGEMENT=y
> 
> IMHO, cache should be user defined since user may want to use ITCM/DTCM for other purposes. So, I'm interested in know how this PR changes the above behavior?

This seems to contradict what I got told in this PR earlier on by @carlocaione. If I understood it correctly SoCs should use in their init code sys_cache_data/instr_enable to enable the caches. Therefore, configuring CONFIG_CACHE_MANAGEMENT=n will result in disabled caches in anyway, and CONFIG_DCACHE and CONFIG_ICACHE will only indicate if the SoC has a cache or not.

Btw, I saw that in the of the PRs you explictly added a disable-call for the caches during initialization. Are the caches enabled using something else earlier on for these SoCs?

carlocaione commented:
> This seems to contradict what I got told in this PR earlier on by @carlocaione. If I understood it correctly SoCs should use in their init code sys_cache_data/instr_enable to enable the caches. Therefore, configuring CONFIG_CACHE_MANAGEMENT=n will result in disabled caches in anyway, and CONFIG_DCACHE and CONFIG_ICACHE will only indicate if the SoC has a cache or not.

yes, this is correct. @nandojve you should never toggle `CONFIG_DCACHE` yourself if you are using the Zephyr cache APIs.

I'll try to better explain what is the rationale here (when using the proper cache APIs):

1. All the zephyr code should use the `sys_cache_*` API to do cache operations so that the code is agnostic about the actual hardware it is running on.
2. The only switch that the user could use to disable / enable the cache operations is `CONFIG_CACHE_MANAGEMENT` and nothing else. If the hardware does not support cache ops that option is doing nothing from the beginning (see following point), otherwise we are simply converting all the cache operations to `-ENOTSUP` when setting that to `=n`.
3. At SoC level the SoC is defining `CONFIG_{D,I}CACHE` whether we support cache operations at SoC level. If the SoC does not support caches, it simply do not define those and the `sys_cache_*` ops are falling to `-ENOTSUP`.

benediktibk commented:
> 1. All the zephyr code should use the `sys_cache_*` API to do cache operations so that the code is agnostic about the actual hardware it is running on.

One small caveat: If external code (for instance HAL) is used for the initalization, which automatically activates caches, the caches have to be deactivated directly after such a call using a low level function call, not sys_cache_*. Then during the SoC initialization later on sys_cache_* can be used to activate the caches.

@carlocaione Correct?

carlocaione commented:
> One small caveat: If external code (for instance HAL) is used for the initalization, which automatically activates caches, the caches have to be deactivated directly after such a call using a low level function call, not sys_cache__. Then during the SoC initialization later on sys_cache__ can be used to activate the caches.
> 
> @carlocaione Correct?

Well yes but I would consider that an HAL issue to be honest. The HAL should not enable the caches behind the user backs, is is up to the OS doing that.

benediktibk committed:
soc: arm: use sys_cache* for enabling the caches in nxp_s32

Use sys_cache* for enabling the caches in nxp_s32. This automatically
considers CONFIG_CACHE_MANAGEMENT and will activate the
cases only if this is active.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
soc: arm: replace redundant config option for caches for nxp_imx

Replace the redundant cache config options for the nxp_imx and
use sys_cache* functions to enable the caches. These will automatically
consider CONFIG_CACHE_MANAGEMENT.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
soc: arm: remove redundant cache config options for kv5x

Remove the redundant cache config options for kv5x and use
the sys_cache* functions to enable the caches. This will automatically
consider CONFIG_CACHE_MANAGEMENT.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
arch: arc: use sys_cache instead of arch-function for enabling the cache

Use sys_cache_data_enable instead of arch_dcache_enable to enable
the cache. This will ensure that CONFIG_CACHE_MANAGEMENT
is considered correctly.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
soc: arm: use sys_cache* for enabling the caches on same70 and samv71

Use the sys_cache* functions to enable the caches on same70 and
samv71. This will ensure that CONFIG_CACHE_MANAGEMENT is
considered correctly.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk commented:
Unfortunately, I'm gonna need some additional information which I could not figure out by myself. For these systems the caches were explicitly disabled during startup. I assume from this fact (and sometimes also a comment was hinting towards this) that a previous call to something like SystemInit, ... automatically enables the caches. We came to the conclusion that it will be necessary to revert these changes directly after this calls.

Could the mentioned people please help me out finding the correct spots?

- soc_rt10xx.c, soc_rt11xx.c, kv5x/soc.c: @dleach02
- same70/soc.c, samv71/soc.c: @nandojve

benediktibk committed:
soc: arm: use sys_cache* to enable caches for stm32f7 and stm32h7

Use sys_cache* functions to enable the caches for stm32f7 and
stm32h7. This ensures that CONFIG_CACHE_MANAGEMENT is
considered correctly.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

nandojve commented:
> Unfortunately, I'm gonna need some additional information which I could not figure out by myself. For these systems the caches were explicitly disabled during startup. I assume from this fact (and sometimes also a comment was hinting towards this) that a previous call to something like SystemInit, ... automatically enables the caches. We came to the conclusion that it will be necessary to revert these changes directly after these calls.
> 
> Could the mentioned people please help me out finding the correct spots?
> 
> * soc_rt10xx.c, soc_rt11xx.c, kv5x/soc.c: @dleach02
> * same70/soc.c, samv71/soc.c: @nandojve

In the Atmel all SoC's are using only one entry point to configure the platform which is _z_arm_platform_init_. If there are any other entry point in the whole initialization that enable/disable cache outside SoC code IMHO that is wrong. After _z_arm_platform_init_ the system should check and use cache if they are available, that is it. It is not possible to the system know everything about the platform, which may include internal flash registers to configure the cache size (for instance). Which means that when customer enables a cache the SRAM size is decreased and that must be considered at build time for this platform (so linker scripts etc). Only the user/customer know that and sometimes the HW need to be programmed to be correct (so it is not only about SW).

The only way that I see to system try configure this is providing a dedicated API, something like _z_arm_platform_cache_XXX. With that, the SoC _may implement_ accordingly the cache initialization. However this could let to other problems like multiple initialization stages.

benediktibk commented:
@nandojve Okay, I think we agree about the one spot where caches should be enabled or disabled and configured correctly. I'm only wondering why there are explicit calls to SCB_DisableI/DCache in z_arm_platform_init at the moment?

    void z_arm_platform_init(void)
    {
        if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_ICACHE)) {
            SCB_EnableICache();
        } else {
            SCB_DisableICache();
        }
        if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_DCACHE)) {
            SCB_EnableDCache();
        } else {
            SCB_DisableDCache();
        }

Are the caches automatically enabled previously, maybe as the default of the hardware?

nandojve commented:
> @nandojve Okay, I think we agree about the one spot where caches should be enabled or disabled and configured correctly. I'm only wondering why there are explicit calls to SCB_DisableI/DCache in z_arm_platform_init at the moment?

They are there because on my perspective the SoC is responsible to do cache enable/disable.

> ```
> void z_arm_platform_init(void)
> {
>     if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_ICACHE)) {
>         SCB_EnableICache();
>     } else {
>         SCB_DisableICache();
>     }
>     if (IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && IS_ENABLED(CONFIG_DCACHE)) {
>         SCB_EnableDCache();
>     } else {
>         SCB_DisableDCache();
>     }
> ```
> 
> Are the caches automatically enabled previously, maybe as the default of the hardware?

The SAME7x/V7x default configuration is that
- DTCM is enabled by default at reset
- ITCM is disabled by default at reset

You can get all details at `11.1.2 Tightly Coupled Memory (TCM) Interface` in SAMV7 datasheet.

nandojve commented on soc/arm/atmel_sam/same70/soc.c:
```diff
@@ -226,16 +227,19 @@ static ALWAYS_INLINE void clock_init(void)
+	 * sys_cache*-functions can enable them, if requested by the
+	 * configuration.
+	 */
+	cache_data_disable();
+
+	/*
+	 * Enable the caches only if configured to do so.
+	 */
+	sys_cache_instr_enable();
+	sys_cache_data_enable();
```
I think this still have issues:

The problem is that the API assumes that CONFIG_CACHE_MANAGEMENT is always enabled and that is not true. If the user say CONFIG_CACHE_MANAGEMENT=NO how the cache is disabled? Maybe it is easy to understand when a bootloader set I-cache TRUE and the system will never disabled because CONFIG_CACHE_MANAGEMENT=NO.

However I think it is safe to replace `SCB_*` by `arch_*` equivalent if that is acceptable.

BTW, I noted that Atmel is not clearing the cache and the system do not clear too for every enable/disable which is a bug.

benediktibk replied:
With this implementation both caches will be disabled if CONFIG_CACHE_MANAGEMENT=N, which is the intended result according to @carlocaione.
z_arm_platform_init will assume a clean start without a bootloader beforehand. For these cases, with a bootloader, I think INIT_ARCH_HW_AT_BOOT should be used, which in turn will trigger a call to INIT_ARCH_HW_AT_BOOT during startup. The implementation there looks faulty to me, as it is using the sys_cache*-functions. But this is a different topic, in my opinion.

nandojve replied:
I agreed that this works only if the INIT_ARCH_HW_AT_BOOT is always enabled for the Atmel SAM after this change. Without that, the initialization can not ensure correct cache initialization if user is using SAM-BA bootloader (ROM bootloader) for instance or any other bootloader due to intrinsic initialization sequence.

So I'll ask you to add the following patch.

```diff
diff --git a/soc/arm/atmel_sam/same70/Kconfig.series b/soc/arm/atmel_sam/same70/Kconfig.series
index 66ceeb6ca3..4e7d6aa396 100644
--- a/soc/arm/atmel_sam/same70/Kconfig.series
+++ b/soc/arm/atmel_sam/same70/Kconfig.series
@@ -14,6 +14,7 @@ config SOC_SERIES_SAME70
        select CPU_HAS_ICACHE
        select CPU_HAS_DCACHE
        select SOC_FAMILY_SAM
+       select INIT_ARCH_HW_AT_BOOT
        select PLATFORM_SPECIFIC_INIT
        select ASF
        select HAS_SWO
diff --git a/soc/arm/atmel_sam/samv71/Kconfig.series b/soc/arm/atmel_sam/samv71/Kconfig.series
index 543def9d9a..cadee35acb 100644
--- a/soc/arm/atmel_sam/samv71/Kconfig.series
+++ b/soc/arm/atmel_sam/samv71/Kconfig.series
@@ -14,6 +14,7 @@ config SOC_SERIES_SAMV71
        select CPU_HAS_ICACHE
        select CPU_HAS_DCACHE
        select SOC_FAMILY_SAM
+       select INIT_ARCH_HW_AT_BOOT
        select PLATFORM_SPECIFIC_INIT
        select ASF
        select HAS_SWO
```

benediktibk replied:
done

nandojve replied:
Hi @benediktibk ,

Thank you for your effort to address this.

benediktibk committed:
arch: arm: core: cortex_m: fix cache disabling in init_arch_hw_at_boot

Use the arch-cache functions instead of the sys-cache-functions
in z_arm_init_arch_hw_at_boot to ensure that the caches are
disabled even when CONFIG_CACHE_MANAGEMENT is disabled.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk commented:
@carlocaione There are, of now, some spots where I have to disable caches, although CONFIG_CACHE_MANAGEMENT might be disabled. I decided to use arch_dcache_disable for this, as I did not want to spread calls to SCB_DisableDCache all over the place. Unfortunately, these functions are only built and linked if CONFIG_CACHE_MANAGEMENT is enabled. To work around this I introduced CONFIG_ANY_CACHE, which is basically the same as CONFIG_CACHE_MANAGEMENT but without a prompt.

Which version would you prefer, CONFIG_ANY_CACHE or calling SCB_DisableDCache directly? Or something completely different?

The problematic spots are:
- arch/arm/core/cortex_m/scb.c
- soc/arm/atmel_sam/same70/soc.c
- soc/arm/atmel_sam/samv71/soc.c

nandojve reviewed (dismissed):
atmel_sam

benediktibk commented:
Removed ANY_CACHE again and used direct calls to SCB_* wherever it was necessary.

benediktibk commented:
Removed ANY_CACHE again and used direct calls to SCB_* wherever it was necessary.

carlocaione reviewed (approved):
Way better IMO

benediktibk commented:
Rebased to trigger CI as the github incident has been resolved.
